akka-javasdk/src/main/scala/akka/javasdk/impl/eventsourcedentity/EventSourcedEntityImpl.scala

```diff
@@ -0,0 +1,295 @@
+import com.google.protobuf.any.{ Any => ScalaPbAny }
+import io.grpc.Status
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.api.trace.Tracer
+import org.slf4j.LoggerFactory
+import org.slf4j.MDC
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[impl] object EventSourcedEntityImpl {
+  private val log = LoggerFactory.getLogger(this.getClass)
+
+  private class CommandContextImpl(
+      override val entityId: String,
+      override val sequenceNumber: Long,
+      override val commandName: String,
+      override val commandId: Long, // FIXME remove
+      override val isDeleted: Boolean,
+      override val metadata: Metadata,
+      span: Option[Span],
+      tracerFactory: () => Tracer)
+      extends AbstractContext
+      with CommandContext
+      with ActivatableContext {
+    override def tracing(): Tracing = new SpanTracingImpl(span, tracerFactory)
+  }
+
+  private class EventSourcedEntityContextImpl(override final val entityId: String)
+      extends AbstractContext
+      with EventSourcedEntityContext
+
+  private final class EventContextImpl(entityId: String, override val sequenceNumber: Long)
+      extends EventSourcedEntityContextImpl(entityId)
+      with EventContext
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[impl] final class EventSourcedEntityImpl[S, E, ES <: EventSourcedEntity[S, E]](
+    configuration: Settings,
+    tracerFactory: () => Tracer,
+    componentId: String,
+    componentClass: Class[_],
+    messageCodec: JsonMessageCodec,
+    factory: EventSourcedEntityContext => ES,
+    snapshotEvery: Int)
+    extends SpiEventSourcedEntity {
+  import EventSourcedEntityImpl._
+
+  if (snapshotEvery < 0)
+    log.warn("Snapshotting disabled for entity [{}], this is not recommended.", componentId)
+
+  // FIXME
+//  private val traceInstrumentation = new TraceInstrumentation(componentId, EventSourcedEntityCategory, tracerFactory)
+
+  private val componentDescriptor = ComponentDescriptor.descriptorFor(componentClass, messageCodec)
+
+  // FIXME remove EventSourcedEntityRouter altogether, and only keep stateless ReflectiveEventSourcedEntityRouter
+  private def createRouter(context: EventSourcedEntityContext)
+      : ReflectiveEventSourcedEntityRouter[AnyRef, AnyRef, EventSourcedEntity[AnyRef, AnyRef]] =
+    new ReflectiveEventSourcedEntityRouter[S, E, ES](
+      factory(context),
+      componentDescriptor.commandHandlers,
+      messageCodec)
+      .asInstanceOf[ReflectiveEventSourcedEntityRouter[AnyRef, AnyRef, EventSourcedEntity[AnyRef, AnyRef]]]
+
+  override def emptyState(entityId: String): SpiEventSourcedEntity.State = {
+    // FIXME rather messy with the contexts here
+    val cmdContext =
+      new CommandContextImpl(entityId, 0L, "", 0, false, MetadataImpl.of(Nil), None, tracerFactory)
+    val context = new EventSourcedEntityContextImpl(entityId)
+    val router = createRouter(context)
+    router.entity._internalSetCommandContext(Optional.of(cmdContext))
```

Comment:
We don't need the command context when accessing emptyState

Reply:
Thanks, that will make this more clean and I can remove the entityId parameter again.

Reply:
maybe we should call it a SpiFactory since not really a descriptor anymore.